New platform media_player/anthemav

nugget opened the pull request:
**Description:**
Add a new media_player platform anthemav to support Anthem MRX and AVM receivers/processors.

Leverages third-party module (also written by submitter) [python-anthemav](https://github.com/nugget/python-anthemav)

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1701

**Example entry for `configuration.yaml` (if applicable):**
```yaml
media_player:
  - platform: anthemav
    host: 10.0.0.42
    port: 14999
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

nugget committed:
Initial commit of anthemav platform.  It loads but has no purpose.

nugget committed:
Now presents a card in the UI but the values aren't real

nugget committed:
Mute and volume polling/setting work now

nugget committed:
Source lists and selection works now.

nugget committed:
Reduce debug logging verbosity

nugget committed:
Support power on/off and skip polling for details if power is off

nugget committed:
Add some static tables to decode numerics from telnet commands

nugget committed:
Add stub for unsupported media_play

nugget committed:
New style anthemav uses native asyncio structure

nugget committed:
Add device callback for asyncio

nugget committed:
This is ugly but it works

nugget committed:
Simplify async setup and abstract class data retrieval

nugget committed:
Implement commands (power on and power off for now)

nugget committed:
Add support for scan_interval and set default to 120 seconds

nugget committed:
Pass-through to package handlers for volume and input selection

nugget committed:
Slight restructuring to satisfy anthemav 0.9

nugget committed:
Merge branch 'dev' into anthemav

nugget committed:
Load anthemav package from pypi now that it's registered

nugget committed:
Proper app_name from a/v info

nugget committed:
Mispelled word

nugget committed:
media_player/anthemav initial commit of platform requirements

scmmmh committed:
Philio 3-in-1 Gen 4 zwave sensor needs the no-off-event workaround. (#5120)

andrey-git committed:
Add print_config_parameter service to Z-Wave (#5121)

* Add print_config_param service to z-wave

* Add print_config_parameter service to z-wave

* Add print_config_parameter service to z-wave

* Fix typos

* Fix typos

* Fix typo

nugget committed:
Merge remote-tracking branch 'upstream/dev' into anthemav

nugget committed:
Conform to Python/project style requirements

nugget committed:
Making pylint happy

nugget committed:
Bring pip requirements in agreement with the code

nugget committed:
Merge remote-tracking branch 'upstream/dev' into anthemav

nugget committed:
Bungled previous update

nugget committed:
Merge remote-tracking branch 'upstream/dev' into anthemav

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,225 @@
+            pval = getattr(self.reader, propname)
+            _LOGGER.debug('query '+propname+' returned from avr: '+str(pval))
+            return pval
+        else:
```
This branch shouldn't ever be necessary.

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,225 @@
+        """Turn AVR power on."""
+        self.update_avr('power', True)
+
+    def volume_up(self):
```
Delete `volume_up` and `volume_down`. They are implemented by `MediaPlayerDevice` if there isn't a platform specific implementation.

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,225 @@
+
+    def mute_volume(self, mute):
+        """Engage AVR mute."""
+        _LOGGER.debug('Request to mute %s', str(mute))
```
Should there be a mute command here? If not remove `SUPPORT_VOLUME_MUTE` and this function.

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,225 @@
+        """Return all active, configured inputs."""
+        return self._poll_and_return('input_list', ["Unknown"])
+
+    def media_play(self):
```
Delete this function.

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,225 @@
+class AnthemAVR(MediaPlayerDevice):
+    """Entity reading values from Anthem AVR protocol."""
+
+    def __init__(self, hass, host, port):
```
Remove `host` and `port` from the constructor and the corresponding attributes if they won't be used.

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,225 @@
+    EVENT_HOMEASSISTANT_STOP, CONF_SCAN_INTERVAL)
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['anthemav>=1.1.3']
```
Mark the requirement as `'anthemav==1.1.3'` for stability. The version can be bumped after testing.

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,225 @@
+            _LOGGER.warning('Unable to issue command to missing AVR')
+
+    @asyncio.coroutine
+    def async_update(self):
```
Delete this and replace with should_poll False. See https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/media_player/cast.py#L112

emlove reviewed (changes requested):
Thanks for you work to expand Home Assistant! Please see the comments I've added to the platform code.

nugget committed:
Remove unnecessady SCAN_INTERVAL logic

I was unawre that this is performed as part of the normal platform behavior and
it's unnecessary for a platform to independently implement this logic.

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,225 @@
+    #
+    # pylint: disable=redefined-outer-name
+    # pylint: disable=invalid-name
+    SCAN_INTERVAL = config.get(CONF_SCAN_INTERVAL) or 120
```
Actually, since the platform will be local push, you can remove `SCAN_INTERVAL` from the module initialization as well.

nugget committed:
Refactor code based on @armills PR requests

nugget committed:
Merge remote-tracking branch 'upstream/dev' into anthemav

nugget committed:
Re-add media_play stub to avoid traceback

nugget committed:
Align with platform reqirements

nugget committed:
Remove references to SCAN_INTERVAL and clean up _lookup logic

nugget committed:
Add DEFAULT_PORT assignment

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,179 @@
+    def __init__(self, hass, avr):
+        """"Initialize entity with hass, host, and port."""
+        super().__init__()
+        self.hass = hass
```
This can be removed from here and the constructor since it's not used.

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,179 @@
+        return getattr(self.reader, propname, dval)
+
+    @property
+    def reader(self):
```
Since this was simplified we might as well just use `self.avr.protocol` instead and eliminate `reader`.

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,179 @@
+    @property
+    def state(self):
+        """Return state of power on/off."""
+        pwrstate = self._lookup('power', None)
```
Just as a code style change we should set the default `dval` in `_lookup` to `None` and remove the argument here.

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,179 @@
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
```
`CONF_NAME` is unused.

armills replied:
This shouldn't be necessary. If the user wants to change the name on the frontend, `customize` can be used. https://home-assistant.io/getting-started/customizing-devices/

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,179 @@
+        """Engage AVR mute."""
+        self.update_avr('mute', mute)
+
+    def update_avr(self, propname, value):
```
We should prefix this function with an underscore as well since it's also for internal use only.

nugget committed:
Code style changes and removal of vestigial structures

nugget committed:
CONF_NAME handling changes to allow local override to default from device

balloob commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,174 @@
+    def anthemav_update_callback(message):
+        """Receive notification from transport that new data exists."""
+        _LOGGER.info('Received update callback from AVR: %s', message)
+        hass.async_add_job(device.async_update_ha_state())
```
You cannot use `async_add_job` when not running inside event loop. Best to replace this whole line with `device.schedule_update_ha_state()`

balloob commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,174 @@
+        self._name = name
+
+    def _lookup(self, propname, dval=None):
+        return getattr(self.avr.protocol, propname, dval)
```
Is this code doing any I/O ?

balloob commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,174 @@
+    def select_source(self, source):
+        """Change AVR to the designated source (by name)."""
+        self._update_avr('input_name', source)
+        return self._lookup('input_list', ["Unknown"])
```
`select_source` does not expect a return value.

nugget committed:
Merge remote-tracking branch 'upstream/dev' into anthemav

nugget committed:
Address PR feedback from @balloob

nugget committed:
Merge remote-tracking branch 'upstream/dev' into anthemav

nugget committed:
Merge remote-tracking branch 'upstream/dev' into anthemav

pvizeli commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,173 @@
+    def anthemav_update_callback(message):
+        """Receive notification from transport that new data exists."""
+        _LOGGER.info('Received update calback from AVR: %s', message)
+        device.schedule_update_ha_state()
```
You have a asyncio library and call all stuff inside Loop. This function can only called from outside the Loop. Use `hass.loop.create_task(device.async_update_ha_state())` but you need sure that you understund asyncio and run all stuff inside Loop!

pvizeli replied:
please rename the function to async_* for represent that will be run inside Loop and use `yield from device.async_update_ha_state()`. Your library is asyncio so I hope this callback run inside Loop.

nugget replied:
Renamed the function for clarity and also removed the media_play stub definition which is no longer needed now that issue #5181 is resolved.

pvizeli commented:
You library is asyncio but you overwrite media_player threaded function. Look on base class and overwrite only `async_*` stuff they will be run inside loop.

nugget committed:
Merge remote-tracking branch 'upstream/dev' into anthemav

nugget committed:
Remove media_play function override

It's no longer necesary for the platform to implement a stub media_play
function override now that the Add SUPPORT_PLAY flag #5181 issue has been
resolved and merged into the dev branch.

nugget committed:
Rename callback function to async_ for clarity

pvizeli commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,169 @@
+    def async_anthemav_update_callback(message):
+        """Receive notification from transport that new data exists."""
+        _LOGGER.info('Received update calback from AVR: %s', message)
+        device.schedule_update_ha_state()
```
use `yield from device.async_update_ha_state()`

pvizeli replied:
Please address also this comment

nugget replied:
I haven't neglected it.  This change breaks the code and I am trying to figure out why.

armills replied:
There's some discussion about the issue here: http://stackoverflow.com/questions/29126340/calling-a-coroutine-from-asyncio-protocol-data-received

Basically the callback needs to be turned into a corouine, and [this line](https://github.com/nugget/python-anthemav/blob/master/anthemav/protocol.py#L279) needs to be able to call a coroutine. `self._loop.create_task(self._update_callback(data))` should work, or if you want to be nice to other users of the library you could have it accept either a coroutine or a normal function by checking with `if asyncio.iscoroutine(self._update_callback):`

nugget replied:
That's extremely helpful, thanks

pvizeli replied:
That is true. you can also use `self.hass.async_add_job(device.async_update_ha_state())` in this case.

pvizeli commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,169 @@
+        """Return all active, configured inputs."""
+        return self._lookup('input_list', ["Unknown"])
+
+    def select_source(self, source):
```
`def async_select_source(...`

pvizeli commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,169 @@
+        """Change AVR to the designated source (by name)."""
+        self._update_avr('input_name', source)
+
+    def turn_off(self):
```
`def async_turn_off(...`

pvizeli commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,169 @@
+        """Turn AVR power off."""
+        self._update_avr('power', False)
+
+    def turn_on(self):
```
same

pvizeli commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,169 @@
+        """Turn AVR power on."""
+        self._update_avr('power', True)
+
+    def set_volume_level(self, volume):
```
same

pvizeli commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,169 @@
+        """Set AVR volume (0 to 1)."""
+        self._update_avr('volume_as_percentage', volume)
+
+    def mute_volume(self, mute):
```
same

pvizeli reviewed (commented):
Looks good. please use async Interface from media_player entity if you library is async.

nugget committed:
Merge remote-tracking branch 'upstream/dev' into anthemav

nugget committed:
Merge remote-tracking branch 'upstream/dev' into anthemav

nugget committed:
Use async routines for platform methods

nugget committed:
Convert update callback to coroutine for conformity

Underlying anthemav library now properly supports coroutine callbacks instead
of normal functions.  Converted the platform callback to a coroutine for
conformance with async operation for the device.

Special thanks to @pvizeli and @armills for their invaluable remedial Python
instruction!

pvizeli commented:
Good work. But one comment :) You are mix 2 comments to one solution. 

1) Use `create_task` on your library and run callback as async.coroutine. Inside this you can use `yield from`
2) Use `call_soon` on your library and run callback as callback. Inside this callback you can use `hass.async_add_job` to execute a coroutine

Way 2 is faster. Please revert your change on your library and use inside the callback `hass.async_add_job(device.async_update_ha_state(True))` instead `yield from device.async_update_ha_state()`. That will work.

Way 1 will process to coro on selector and that will be slowdone.

nugget commented:
I'm afraid I do not understand what you are suggesting, sorry.  "Way 1" is code I want to avoid?  Or is that part of the approach you prefer?

Do you want me to replace the `hass.async_add_job` in the callback function with a `yield from`?  And the "Way 1" code will allow that to work?  Where does `call_soon` belong?  In the library or in the platform code?

armills commented:
What he's saying is the advice I gave works, but it's not the optimal asyncio solution. You want to keep the callback logic in the platform the same, but get rid of the coroutine decorator. You then want to change [this line](https://github.com/nugget/python-anthemav/blob/master/anthemav/protocol.py#L289) to `self._loop.call_soon(self._update_callback, data)`. The coroutine branch won't be used by hass, so it could be removed, or you can keep it if you want to keep the support in your library.

nugget committed:
Further callback refinements

Altered the nature of callback handling based on suggestions from @pvizeli

armills commented on homeassistant/components/media_player/anthemav.py:
```diff
@@ -0,0 +1,174 @@
+    def async_anthemav_update_callback(message):
+        """Receive notification from transport that new data exists."""
+        _LOGGER.info('Received update calback from AVR: %s', message)
+        hass.async_add_job(device.async_update_ha_state(True))
```
The True shouldn't be necessary since this is local push and doesn't have an update method.

nugget committed:
Merge remote-tracking branch 'upstream/dev' into anthemav

nugget committed:
True not needed for local push update_ha_state

pvizeli committed:
Small style fix

pvizeli commented:
Good work ⚡️

armills commented:
Thanks for putting up with us!
